### PR TITLE
Support multiple parameter groups in LR scheduler factories

### LINKED ISSUE #1316: Support multiple parameter groups in recipe LR factories

### PR DESCRIPTION
**Pull request by cbalioglu**
This PR (1) adds support for multiple parameter groups in LR scheduler factories, (2) starts adding/improving docstrs for existing and new code constructs, and (3) adds unit tests for all LR factories with near 100% coverage. Resolves #1316

### CODE REVIEW

File: doc/source/reference/api/index.rst [hunk 1]
```diff
@@ -58,3 +58,4 @@ Utilities
    device
    logging
    metrics
```
Comment (cbalioglu) [inline doc/source/reference/api/index.rst:60]: I think we should move to a package-based file naming convention for the rest of the reference docs.

File: src/fairseq2/optim/lr_schedulers/cosine_annealing.py [hunk 2]
```diff
@@ -121,31 +116,31 @@ def _compute_lrs(self) -> list[float]:
...
-        lr_mul = self._lr_mul**cycle_nr
+        lr_mul = self.lr_mul**cycle_nr
 
         c = math.cos(math.pi * cycle_pos / cycle_len)
 
-        min_lrs, max_lrs = self._final_lrs, base_lrs
+        min_lrs, max_lrs = self.final_lrs, base_lrs
 
-        return [self._cycle_lr(mn, mx, lr_mul, c) for mn, mx in zip(min_lrs, max_lrs)]
+        def cycle_lr(min_lr: float, max_lr: float) -> float:
```
Comment (zyaoj) [inline src/fairseq2/optim/lr_schedulers/cosine_annealing.py:140]: Curious about the motivation of moving `_cycle_lr` from a private method to a local function.

File: src/fairseq2/recipe/config.py [hunk 3]
```diff
@@ -908,16 +940,25 @@ def validate(self) -> ValidationResult:
 
 @dataclass(kw_only=True)
 class TriStageLRConfig(Validatable):
+    """Represents the configuration of :class:`TriStageLRConfig`."""
```
Comment (zyaoj) [inline src/fairseq2/recipe/config.py]: Did you want to put:
```suggestion
    """Represents the configuration of :class:`TriStageLR`."""
```

File: src/fairseq2/recipe/trainer.py [hunk 4]
```diff
@@ -881,7 +883,10 @@ def _publish_metrics(self) -> None:
 
             self._unit.process_metric_values(values)
 
-            values["lr"] = self._last_lr
+            if len(self._last_lrs) == 1:
```
Comment (zyaoj) [inline src/fairseq2/recipe/trainer.py:888]: [nit] Feels like this could be potentially confusing for people who are not aware of our code and the goal of supporting multiple lr. Maybe add a comment here or abstract these for lines into a helper function?

File: src/fairseq2/recipe/optim.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            if isinstance(start_lr, float):
+                return [start_lr] * num_param_groups
+
+            maybe_raise_param_group_length_error("start_lr", start_lr, num_param_groups)
+
+            return start_lr
+    """
+    if len(value) != num_param_groups:
+        raise ValidationError(
+            f"The length of `{field}` must match the number of optimizer parameter groups ({num_param_groups}), but is {len(value)} instead.", field="lr_scheduler.config"  # fmt: skip
```
Comment (zyaoj) [inline src/fairseq2/recipe/optim.py:56]: [nit] Make error message consistent across the board (or merge into the same function?) -- e.g. in [`optim.lr_schedulers.lr_scheduler`](https://github.com/facebookresearch/fairseq2/blob/main/src/fairseq2/optim/lr_schedulers/lr_scheduler.py#L57):

```py
    if len(value) != num_param_groups:
        raise ValueError(
            f"The length of `{name}` must be equal to the number of parameter groups ({num_param_groups}), but is {len(value)} instead."
        )
```

File: doc/source/reference/api/index.rst (hunk 1 above)
Comment (zyaoj) [reply doc/source/reference/api/index.rst:60]: Yeah many of those files were placeholders. Do we want to use a simple flat structure then instead of the hierarchical layout?

**Review by zyaoj — APPROVED** [review]
LGTM! Just some nit comment and questions. Thank you for boosting the pytest coverage!

```
src/fairseq2/optim/__init__.py                                   3      0   100%
src/fairseq2/optim/fp16_loss_scaler.py                         130     62    52%
src/fairseq2/optim/lr_schedulers/__init__.py                    10      0   100%
src/fairseq2/optim/lr_schedulers/cosine_annealing.py            43      0   100%
src/fairseq2/optim/lr_schedulers/lr_scheduler.py                34      4    88%
src/fairseq2/optim/lr_schedulers/myle.py                        22      1    95%
src/fairseq2/optim/lr_schedulers/noam.py                        18      0   100%
src/fairseq2/optim/lr_schedulers/polynomial_decay.py            31      1    97%
src/fairseq2/optim/lr_schedulers/tri_stage.py                   42      1    98%
src/fairseq2/recipe/internal/lr_schedulers.py                  122      9    93%

File: doc/source/reference/api/index.rst (hunk 1 above)
Comment (cbalioglu) [reply doc/source/reference/api/index.rst:60]: Yes, I think that will be easier to maintain as well as to quickly spot which file maps to which module.

File: src/fairseq2/recipe/config.py (hunk 3 above)
Comment (cbalioglu) [reply src/fairseq2/recipe/config.py]: Oh good catch, thanks!

File: src/fairseq2/optim/lr_schedulers/cosine_annealing.py (hunk 2 above)
Comment (cbalioglu) [reply src/fairseq2/optim/lr_schedulers/cosine_annealing.py:140]: `_cycle_len` was only used in `_compute_lrs` as part of the last list comprehension statemet and required capturing `lr_mul` and `c` local variables. I thought that it would be cleaner and also reduce code if it became a local function (two less parameters to pass).

File: doc/source/reference/api/index.rst (hunk 1 above)
Comment (zyaoj) [reply doc/source/reference/api/index.rst:60]: Sounds good. Will create a follow-up PR to restructure. Also a quick note that there are many examples squeezed into the reference api pages for now. We should probably also surface them to basics or tutorials.
